src/monodroid/jni/timing.hh

```diff
@@ -0,0 +1,161 @@
+#ifndef __TIMING_HH
+#define __TIMING_HH
+
+#include <stdarg.h>
+#include <cstdlib>
+#include <sys/time.h>
+
+#ifdef ANDROID
+#include <android/log.h>
+#endif
+
+#include "cppcompat.hh"
+#include "logger.hh"
+
+namespace xamarin::android
+{
+	struct timing_point
+	{
+		time_t sec;
+		uint64_t ns;
+
+		void mark ();
+	};
+
+	struct timing_period
+	{
+		timing_point start;
+		timing_point end;
+
+		void mark_start ()
+		{
+			start.mark ();
+		}
+
+		void mark_end ()
+		{
+			end.mark ();
+		}
+	};
+
+	struct timing_diff
+	{
+		static constexpr uint32_t ms_in_nsec = 1000000ULL;
+
+		time_t sec;
+		uint32_t ms;
+		uint32_t ns;
+
+		timing_diff (const timing_period &period);
+	};
+
+	struct managed_timing_sequence
+	{
+		timing_period  period;
+		bool           in_use;
+		bool           dynamic;
+	};
+
+	// This class is intended to be used by the managed code. It can be used by the native code as
+	// well, but the overhead it has (out of necessity) might not be desirable in native code.
+#define TIMING_FORMAT "; elapsed: %lis:%lu::%lu"
+
+	class Timing
+	{
+		static constexpr char MESSAGE_FORMAT[] = "%s" TIMING_FORMAT;
+
+	public:
+		static constexpr size_t DEFAULT_POOL_SIZE = 16;
+
+	public:
+		explicit Timing (size_t initial_pool_size = DEFAULT_POOL_SIZE) noexcept
+			: sequence_pool_size (initial_pool_size)
+		{
+			sequence_pool = new managed_timing_sequence [initial_pool_size] ();
+		}
+
+		~Timing () noexcept
+		{
+			delete[] sequence_pool;
+		}
+
+		static void debug (timing_period const &period, const char *message) noexcept
+		{
+			timing_diff diff (period);
+
+			log_debug_nocheck (LOG_TIMING, MESSAGE_FORMAT, message == nullptr ? "" : message, diff.sec, diff.ms, diff.ns);
+		}
+
+		static void info (timing_period const &period, const char *message) noexcept
+		{
+			timing_diff diff (period);
+
+			log_info_nocheck (LOG_TIMING, MESSAGE_FORMAT, message == nullptr ? "" : message, diff.sec, diff.ms, diff.ns);
```

Comment:
~Same question as with `Timing::debug()`: why the `_nocheck` variant?  *Currently* all callers appear to check the log level before calling `Timing::info` before hand, but will that remain the case in the future?

Reply:
Yes, this is the intention - that it will remain so in the future.